File: agc/AGC033/a.cpp
#include <string>
#include <iostream>
#include <cstdio>
#include <queue>
#include <tuple>

using namespace std;

char Black = '#';
char White = '.';
int dir_x[4] = {1, 0, -1, 0};
int dir_y[4] = {0, 1, 0, -1};

int main(int argc, char const *argv[])
{
    int H = 0;
    int W = 0;
    int step = 0;
    if (scanf("%d %d", &H, &W) < 2)
    {
        fprintf(stderr, "error: ill-formed fraction.\n");
        exit(1);
    }
    int memo[1001][1001] = {};
    int count = H * W;
    queue<tuple<int, int>> q;
    // printf("debug");
    for (int y = 0; y < H; y++)
    {
        string temp;
        cin >> temp;
        for (int x = 0; x < W; x++)
        {
            if (temp[x] == Black)
            {
                memo[y][x] = 0;
                q.push(make_tuple(y, x));
                count--;
            }
            else
            {
                memo[y][x] = -1;
            }
        }
    }
    // printf("debug");
    while (count > 0)
    {
        step++;
        queue<tuple<int, int>> next_queue;
        while (!q.empty())
        {
            tuple<int, int> temp_tuple = q.front();
            q.pop();
            int ly = get<0>(temp_tuple);
            int lx = get<1>(temp_tuple);
            for (int i = 0; i < 4; i++)
            {
                int temp_x = lx + dir_x[i];
                int temp_y = ly + dir_y[i];
                if (temp_x >= 0 && temp_x < W && temp_y >= 0 && temp_y < H)
                {
                    if (memo[temp_y][temp_x] < 0)
                    {
                        memo[temp_y][temp_x] = step;
                        count -= 1;
                        next_queue.push(make_tuple(temp_y, temp_x));
                    }
                }
            }
        }
        q = next_queue;
    }

    printf("%d\n", step);
    return 0;
}

File: agc/AGC033/a.py
# coding:utf-8

import copy

Black = "#"
White = "."

dir_x = [1, 0, -1, 0]
dir_y = [0, 1, 0, -1]


def main():
    H, W = map(int, input().split())
    masu = []
    for i in range(H):
        masu.append(input().rstrip())
    step = 0
    count = W*H
    memo = [[-1]*W for i in range(H)]

    steplist = []
    for y in range(H):
        for x in range(W):
            if masu[y][x] == Black:
                memo[y][x] = 0
                count -= 1
                steplist.append([y, x])

    while count > 0:
        temp_step = step
        step += 1
        next_list = []
        for l in steplist:
            y = l[0]
            x = l[1]
            for i in range(4):
                temp_x = x + dir_x[i]
                temp_y = y + dir_y[i]
                if temp_x >= 0 and temp_x < W and temp_y >= 0 and temp_y < H and memo[temp_y][temp_x] < 0:
                    memo[temp_y][temp_x] = step
                    count -= 1
                    next_list.append([temp_y, temp_x])
        steplist = copy.deepcopy(next_list)
    print(step)


if __name__ == "__main__":
    main()
